// @ton-community/assets-sdk/dist/common/build-contract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildContract = void 0;
const path_1 = require("path");
const func_js_1 = require("@ton-community/func-js");
function pascalCase(str) {
    return `${str}`
        .replace(/^(.)/, (_, firstLetter) => firstLetter.toLowerCase())
        .replace(/[-_](.)/g, (_, nextLetter) => nextLetter.toUpperCase());
}
function fileName(src) {
    return (0, path_1.parse)(src).name;
}
function generateParamName(output) {
    return pascalCase(fileName(output)) + 'Code';
}
async function buildContract(args) {
    const compileResult = await (0, func_js_1.compileFunc)({
        sources: args.sources,
        targets: args.targets,
    });
    if (compileResult.status === 'error') {
        throw new Error(compileResult.message);
    }
    const { codeBoc } = compileResult;
    const paramName = generateParamName(args.output);
    const content = `export const ${paramName} = {
        codeBoc: '${codeBoc}'
    };`;
    const { writer } = args;
    writer(args.output, Buffer.from(content, 'ascii'));
    console.log(`Contract ${args.output} was successfully built`);
}
exports.buildContract = buildContract;
