// @ton-community/assets-sdk/dist/jetton/contracts/compile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const reader_1 = require("../../common/reader");
const build_contract_1 = require("../../common/build-contract");
const writer_1 = require("../../common/writer");
// func -SPA -o ./build/jetton-minter-discoverable.fif ../stdlib.fc params.fc op-codes.fc discovery-params.fc jetton-utils.fc jetton-minter-discoverable.fc
async function buildJettonMasterContract() {
    return (0, build_contract_1.buildContract)({
        sources: (0, reader_1.reader)(__dirname),
        writer: (0, writer_1.writer)(__dirname),
        targets: [
            '../../common/stdlib.fc',
            './params.fc',
            './op-codes.fc',
            './discovery-params.fc',
            './jetton-utils.fc',
            './jetton-minter-discoverable.fc',
        ],
        output: './build/jetton-minter.ts',
    });
}
// func -SPA -o ./build/jetton-wallet.fif ../stdlib.fc params.fc op-codes.fc jetton-utils.fc jetton-wallet.fc
async function buildJettonWalletContract() {
    return (0, build_contract_1.buildContract)({
        sources: (0, reader_1.reader)(__dirname),
        writer: (0, writer_1.writer)(__dirname),
        targets: [
            '../../common/stdlib.fc',
            './params.fc',
            './op-codes.fc',
            './jetton-utils.fc',
            './jetton-wallet.fc',
        ],
        output: './build/jetton-wallet.ts',
    });
}
async function main() {
    await buildJettonMasterContract();
    await buildJettonWalletContract();
}
main().catch((err) => {
    throw err;
});
