// @ton-community/assets-sdk/dist/cli/get-nft-item.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = void 0;
const common_1 = require("./common");
const core_1 = require("@ton/core");
const inquirer_1 = __importDefault(require("inquirer"));
async function promptForUserInput() {
    const { item } = await inquirer_1.default.prompt([{
            name: 'item',
            message: 'Enter NFT item address',
        }]);
    return {
        nftAddress: core_1.Address.parse(item),
    };
}
async function main() {
    const { sdk, network } = await (0, common_1.createEnv)();
    const { nftAddress } = await promptForUserInput();
    const nft = sdk.openNftItem(nftAddress);
    const nftData = await nft.getData();
    const nftContent = await nft.getContent();
    const itemInfo = {
        name: nftContent.name,
        description: nftContent.description,
        image: nftContent.image?.toString('base64'),
        owner: nftData.owner,
        collection: nftData.collection,
        index: nftData.index,
    };
    (0, common_1.printInfo)(itemInfo, network);
}
exports.main = main;
