// @ton-community/assets-sdk/dist/cli/get-wallet-state.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = void 0;
const common_1 = require("./common");
const core_1 = require("@ton/core");
const inquirer_1 = __importDefault(require("inquirer"));
async function promptForUserInput(params) {
    const { address } = await inquirer_1.default.prompt([{
            name: 'address',
            message: 'Enter wallet address (default: your wallet address)',
            default: params.defaultOwner,
        }]);
    return {
        address: core_1.Address.parse(address),
    };
}
async function main() {
    const { client, sender, network } = await (0, common_1.createEnv)();
    const { address } = await promptForUserInput({
        defaultOwner: (0, common_1.formatAddress)(sender.address, network)
    });
    const account = await client.provider(address).getState();
    const walletInfo = {
        address: address,
        type: account.state.type,
        balance: (0, core_1.fromNano)(account.balance) + ' TON',
    };
    (0, common_1.printInfo)(walletInfo, network);
}
exports.main = main;
