// @ton-community/assets-sdk/dist/cli/get-jetton-balance.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = void 0;
const common_1 = require("./common");
const core_1 = require("@ton/core");
const inquirer_1 = __importDefault(require("inquirer"));
async function promptForUserInput(params) {
    const { jetton, owner } = await inquirer_1.default.prompt([{
            name: 'jetton',
            message: 'Enter jetton address',
        }, {
            name: 'owner',
            message: 'Enter owner address (default: your wallet address)',
            default: params.defaultOwner,
        }]);
    return {
        jettonAddress: core_1.Address.parse(jetton),
        ownerAddress: core_1.Address.parse(owner),
    };
}
async function main() {
    const { sdk, network, sender } = await (0, common_1.createEnv)();
    const { jettonAddress, ownerAddress } = await promptForUserInput({
        defaultOwner: (0, common_1.formatAddress)(sender.address, network)
    });
    const jetton = sdk.openJetton(jettonAddress);
    const jettonWallet = await jetton.getWallet(ownerAddress);
    const jettonWalletData = await jettonWallet.getData();
    const jettonWalletInfo = {
        name: 'Jetton Wallet',
        jetton: jettonWalletData.jettonMaster,
        'jetton wallet': jettonWallet.address,
        owner: jettonWalletData.owner,
        balance: jettonWalletData.balance,
    };
    (0, common_1.printInfo)(jettonWalletInfo, network);
}
exports.main = main;
