// @ton-community/assets-sdk/dist/cli/get-jetton.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = void 0;
const common_1 = require("./common");
const core_1 = require("@ton/core");
const inquirer_1 = __importDefault(require("inquirer"));
async function promptForUserInput() {
    const { jetton } = await inquirer_1.default.prompt([{
            name: 'jetton',
            message: 'Enter jetton address',
        }]);
    return {
        jettonAddress: core_1.Address.parse(jetton),
    };
}
async function main() {
    const { sdk, network } = await (0, common_1.createEnv)();
    const { jettonAddress } = await promptForUserInput();
    const jetton = sdk.openJetton(jettonAddress);
    const jettonData = await jetton.getData();
    const jettonContent = await jetton.getContent();
    const jettonInfo = {
        name: jettonContent.name,
        description: jettonContent.description,
        image: jettonContent.image?.toString('base64'),
        totalSupply: jettonData.totalSupply,
        mintable: jettonData.mintable,
        adminAddress: jettonData.adminAddress,
        symbol: jettonContent.symbol,
        decimals: jettonContent.decimals,
    };
    (0, common_1.printInfo)(jettonInfo, network);
}
exports.main = main;
