// @ton-community/assets-sdk/dist/cli/get-nft-collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = void 0;
const common_1 = require("./common");
const core_1 = require("@ton/core");
const inquirer_1 = __importDefault(require("inquirer"));
async function promptForUserInput() {
    const { collection } = await inquirer_1.default.prompt([{
            name: 'collection',
            message: 'Enter collection address',
        }]);
    return {
        collectionAddress: core_1.Address.parse(collection),
    };
}
async function main() {
    const { sdk, network } = await (0, common_1.createEnv)();
    const { collectionAddress } = await promptForUserInput();
    const collection = sdk.openNftCollection(collectionAddress);
    const collectionData = await collection.getData();
    const collectionContent = await collection.getContent();
    const collectionInfo = {
        name: collectionContent.name,
        description: collectionContent.description,
        image: collectionContent.image?.toString('base64'),
        owner: collectionData.owner,
        items: collectionData.nextItemIndex,
    };
    (0, common_1.printInfo)(collectionInfo, network);
}
exports.main = main;
