// @ton-community/assets-sdk/dist/cli/transfer-nft.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = void 0;
const core_1 = require("@ton/core");
const common_1 = require("./common");
const inquirer_1 = __importDefault(require("inquirer"));
async function promptForUserInput() {
    const { item, recipient } = await inquirer_1.default.prompt([{
            name: 'item',
            message: 'Enter item address',
        }, {
            name: 'recipient',
            message: 'Enter recipient address',
        }]);
    return {
        nftAddress: core_1.Address.parse(item),
        recipient: core_1.Address.parse(recipient),
    };
}
async function main() {
    const { sdk, network, sender } = await (0, common_1.createEnv)();
    const { nftAddress, recipient } = await promptForUserInput();
    const item = sdk.openNftItem(nftAddress);
    await item.send(sender, recipient);
    const nftTransferInfo = {
        name: 'Transfer NFT',
        item: item.address,
        recipient: recipient,
    };
    (0, common_1.printInfo)(nftTransferInfo, network);
}
exports.main = main;
