// @ton-community/assets-sdk/dist/cli/transfer-ton.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = void 0;
const core_1 = require("@ton/core");
const common_1 = require("./common");
const inquirer_1 = __importDefault(require("inquirer"));
async function promptTonTransfer() {
    const { recipient, amount } = await inquirer_1.default.prompt([{
            name: 'recipient',
            message: 'Enter recipient address',
        }, {
            name: 'amount',
            type: 'number',
            message: 'Enter amount in TON (eg. 1.5)',
            validate: (input) => {
                const amount = (0, core_1.toNano)(input);
                return amount > 0 ? true : 'Price must be a positive number';
            },
        }]);
    return {
        recipient: core_1.Address.parse(recipient),
        amount: (0, core_1.toNano)(amount),
    };
}
async function main() {
    const { sdk, network } = await (0, common_1.createEnv)();
    const { recipient, amount } = await promptTonTransfer();
    await sdk.sender?.send({ to: recipient, value: amount });
    const tonTransferInfo = {
        name: 'Transfer TON',
        recipient: recipient,
        amount: (0, core_1.fromNano)(amount) + ' TON',
    };
    (0, common_1.printInfo)(tonTransferInfo, network);
}
exports.main = main;
