// @ton-community/assets-sdk/dist/nft/contracts/compile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const reader_1 = require("../../common/reader");
const build_contract_1 = require("../../common/build-contract");
const writer_1 = require("../../common/writer");
// func -o build/nft-collection-editable-code.fif -SPA ../stdlib.fc params.fc op-codes.fc nft-collection-editable.fc
async function buildNftCollectionEditableContract() {
    return (0, build_contract_1.buildContract)({
        sources: (0, reader_1.reader)(__dirname),
        writer: (0, writer_1.writer)(__dirname),
        targets: [
            '../../common/stdlib.fc',
            './params.fc',
            './op-codes.fc',
            './nft-collection-editable.fc',
        ],
        output: './build/nft-collection-editable.ts',
    });
}
// func -o build/nft-item-code.fif -SPA ../stdlib.fc params.fc op-codes.fc nft-item.fc
async function buildNftItemContract() {
    return (0, build_contract_1.buildContract)({
        sources: (0, reader_1.reader)(__dirname),
        writer: (0, writer_1.writer)(__dirname),
        targets: [
            '../../common/stdlib.fc',
            './params.fc',
            './op-codes.fc',
            './nft-item.fc',
        ],
        output: './build/nft-item.ts',
    });
}
// func -o build/sbt-item-code.fif -SPA ../stdlib.fc params.fc op-codes.fc sbt-item.fc
async function buildSbtItemContract() {
    return (0, build_contract_1.buildContract)({
        sources: (0, reader_1.reader)(__dirname),
        writer: (0, writer_1.writer)(__dirname),
        targets: [
            '../../common/stdlib.fc',
            './params.fc',
            './op-codes.fc',
            './sbt-item.fc',
        ],
        output: './build/sbt-item.ts',
    });
}
async function main() {
    await buildNftCollectionEditableContract();
    await buildNftItemContract();
    await buildSbtItemContract();
}
main().catch((err) => {
    throw err;
});
